### PR TITLE
Use poetry to lock dependencies

### PR DESCRIPTION
**Pull request by DMRobertson**
This set of changes
- ditches setup.py in favour of poetry configuration in pyproject.toml
- `.gitignore`s poetry's lockfile. (`signedjson` is a library---there is no fixed set of dependencies it ships with and we should not allow ourselves the luxury of thinking so!)
- notes that `setup.py build` can be replaced with `poetry build`.

From the bigger picture, this is a practice run to let us
- experience poetry in use
- (in subsequent PRs) work out how to use poetry in CI

## How to try this out

1. Install poetry.
 - They [recommend](https://python-poetry.org/docs/#installation) using a script to install a self-managed version of poetry in `~/.poetry`.
 - I naively used `pip install --user poetry`.
 - I've seen others recommend [`pipx install poetry`](https://github.com/pypa/pipx).
 - Check it's on your path: `which poetry`.
2. checkout and `cd` to this repository.
3. `poetry install`. This will create a virtualenv behind the scenes and populate it with the right packages.
4. To actually use the virtualenv:
  - `poetry env info` shows you which interpreter you're running, and the virtualenv location
  - for one-off commands: `poetry run echo hello`
  - to activate the virtualenv in a new shell: `poetry shell`. `exit` or Ctrl-D effectively quits the venv, instead of `deactivate`.
  - See `poetry env --help` for details of how you can switch between different environments.

### direnv

I found it really pleasant to use [`direnv`](https://direnv.net/) to automatically switch to/from the env as I navigated into/out of the directory. Needed some extra config to support poetry though, see [here](https://rgoswami.me/posts/poetry-direnv/) (which eventually made its way into the [wiki](https://github.com/direnv/direnv/wiki/Python#poetry)). By the looks of it, this packages queries poetry to find out where the current venv is on disk, and appends its `bin` directory to your path. This doesn't seem to detect use of `poetry env` to switch to another virtual environment within the project.

## What do the built packages look like?

We want to convince ourselves that poetry generates and equivalent sdist to setuptools. One can test as follows:

1. `poetry build`. A wheel and sdist should be generated in `dist`.
1. `pushd dist; tar -xzf signedjson-1.1.1.tar.gz; popd;`
1. `curl https://files.pythonhosted.org/packages/dc/f2/2b3c5574ab77e086597cdc85781fa1d2ef1b4e55bd53d47308bbd2ad7596/signedjson-1.1.1.tar.gz > /tmp/signedjson-pypi.tar.gz`
1. `pushd /tmp; tar -zxf signedjson-pypi.tar.gz; popd`
1. `diff -ru /tmp/signedjson-1.1.1/ dist/signedjson-1.1.1  > diff.txt`

Here's my [diff.txt](https://github.com/matrix-org/python-signedjson/files/7915908/diff.txt).

It's probably most instructive to see which files have been added and removed as follows:

```shell
$ tree /tmp/signedjson-1.1.1/ > pypi.txt
$ tree dist/signedjson-1.1.1/ > poetry.txt
$ diff -u pypi.txt poetry.txt 
```

This yields
```diff
--- pypi.txt	2022-01-21 18:45:03.976920722 +0000
+++ poetry.txt	2022-01-21 18:45:11.595932982 +0000
@@ -1,24 +1,19 @@
-/tmp/signedjson-1.1.1/
+dist/signedjson-1.1.1/
 ├── changelog.d
+│   ├── 11.misc
+│   ├── 18.misc
+│   └── 9.bugfix
 ├── CHANGELOG.md
 ├── LICENSE
-├── MANIFEST.in
 ├── PKG-INFO
 ├── pyproject.toml
 ├── README.rst
-├── setup.cfg
 ├── setup.py
 ├── signedjson
 │   ├── __init__.py
 │   ├── key.py
 │   ├── sign.py
 │   └── types.py
-├── signedjson.egg-info
-│   ├── dependency_links.txt
-│   ├── PKG-INFO
-│   ├── requires.txt
-│   ├── SOURCES.txt
-│   └── top_level.txt
 ├── tests
 │   ├── __init__.py
 │   ├── test_key.py
@@ -26,4 +21,4 @@
 │   └── test_sign.py
 └── tox.ini
 
-4 directories, 22 files
+3 directories, 18 files
```

### CODE REVIEW

**Comment by callahad** [comment]
Would you mind throwing in a recursive diff of the two sdists (poetry and current)?

File: .gitignore [hunk 1]
```diff
@@ -7,3 +7,6 @@ __pycache__
 */*.pyc
 build/
 dist/
+
+# For direnv users.
```
Comment (clokep) [inline .gitignore:11]: Should we gitignore `.idea` and remove those files?

Comment (DMRobertson) [reply .gitignore:11]: That's what I'd do, but maybe @richvdh  had a good reason in https://github.com/matrix-org/python-signedjson/commit/acdb658ad992e6b3bd9a1494d8ec307399bd0830?

**Comment by DMRobertson** [comment]
> Would you mind throwing in a recursive diff of the two sdists (poetry and current)?

Description updated with a diff between `tree` on the two sdists.

**Comment by DMRobertson** [comment]
> Description updated with a diff between `tree` on the two sdists.

Note that
-  poetry seems to _generate_ its own `setup.py` in the sdist (a compatibility aid?). It does not generate a `setup.cfg`
- setuptools generated a very boring `setup.cfg`:
  ```$ cat FROM_PYPI/signedjson-1.1.1/setup.cfg
  [egg_info]
  tag_build = 
  tag_date = 0
  ```

Additionally, the contents of the (new) wheel are minimal:

```
 tree -a wheel
wheel
├── signedjson
│   ├── __init__.py
│   ├── key.py
│   ├── sign.py
│   └── types.py
├── signedjson-1.1.1.dist-info
│   ├── LICENSE
│   ├── METADATA
│   ├── RECORD
│   └── WHEEL
└── signedjson-1.1.1-py3-none-any.whl

2 directories, 9 files
```

**Comment by DMRobertson** [comment]
For cross referencing, https://github.com/matrix-org/synapse/issues/11537

File: .gitignore (hunk 1 above)
Comment (richvdh) [reply .gitignore:11]: there's stuff in `.idea` that is meant to be shared. See https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems, and also https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore which is github's suggestion for a `.gitignore` file

Comment (callahad) [reply .gitignore:11]: s/meant to be shared/can be shared/ ;)

Let's tackle what to do with .idea in a separate issue (and thus remove the changes to .idea/* from this PR, unless critical)

**Comment by callahad** [comment]
> * poetry seems to _generate_ its own `setup.py` in the sdist (a compatibility aid?). It does not generate a `setup.cfg`

Yep. Poetry 1.2 will stop generating setuptools stubs in sdists by default, now that support for PEP 517 builds is widespread. There's a `generate-setup-file` flag which can be used to skip generating the file in older Poetry versions, but it's fine to leave it as-is.

File: .gitignore (hunk 1 above)
Comment (DMRobertson) [reply .gitignore:11]: This is 7ef8b8f.

**Comment by DMRobertson** [comment]
>> Would you mind throwing in a recursive diff of the two sdists (poetry and current)?
> 
> Description updated with a diff between `tree` on the two sdists.

Oh, I've just learned that `diff -r` is a thing. Description updated.

**Comment by DMRobertson** [comment]
I'm going to force push a simpler commit history. Apologies all, but I think it's more helpful than to see my works in progress.

**Comment by DMRobertson** [comment]
This was a good exercise, but I think sydent (being an application, not a library) makes for a better test bed.

**Pull request closed without merging**